genie-web/src/main/java/com/netflix/genie/web/util/StreamBuffer.java

```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ *  Copyright 2019 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.web.util;
+
+import com.google.protobuf.ByteString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.NotImplementedException;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A temporary in-memory structure to hold in-transit data.
+ * Provides an {@code InputStream} for reading, reading blocks until data becomes available or the buffer is closed.
+ * <p>
+ * To avoid in-memory data growing excessively, this buffer stores a single "chunk" at the time.
+ * Only after a chunk is consumed, a new one can be appended.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@ThreadSafe
+@Slf4j
+public class StreamBuffer {
+
+    private final Object lock = new Object();
+    private final AtomicReference<StreamBufferInputStream> inputStreamRef = new AtomicReference<>();
+
+    private boolean closed;
+    private ByteString currentChunk;
+    private int currentChunkWatermark;
+
+    /**
+     * Constructor.
+     */
+    public StreamBuffer() {
+        this.inputStreamRef.set(new StreamBufferInputStream(this));
+    }
+
+    /**
+     * Close this buffer before all data is written due to an error.
+     * Reading will return the end of stream marker after the current chunk (if any) has been consumed.
+     *
+     * @param t the cause for the buffer to be closed.
+     */
+    public void closeForError(final Throwable t) {
+        log.error("Closing buffer due to error: " + t.getClass().getSimpleName() + ": " + t.getMessage());
+        this.closeForCompleted();
+    }
+
+    /**
+     * Close this buffer because all expected data has been written
+     * Reading will return the end of stream marker after all data has been consumed.
+     */
+    public void closeForCompleted() {
+        synchronized (this.lock) {
+            this.closed = true;
+            this.lock.notifyAll();
+        }
+    }
+
+    /**
+     * Append a chunk of data for consumption.
+     * This call may block and not return until some data is read/consumed.
+     *
+     * @param data the data to write into the buffer
+     * @throws IllegalStateException if writing is attempted after the buffer has been closed
+     */
+    public void write(final ByteString data) {
```

Comment:
The use of `ByteString` here makes this pretty much a protobuf/grpc only utility. Could we just unwrap the `ByteString` into a byte array and make it more general?

Reply:
I could at the cost of one extra copy. Do you think it's worth it?